### PR TITLE
Refactoring currency context & related code

### PR DESCRIPTION
**Pull request by pamil**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | yes
| BC breaks?      | yes
| Related tickets | #5488
| License         | MIT

**DONE:**
- [x] Remove currency from `sylius_general` settings
- [x] Remove currency from `Customer` model
- [x] Decouple CurrencyBundle from MoneyBundle
- [x] Make Currency-Money Bundles integration optional
- [x] Reduce amount of different helpers
- [x] Make forms independent from contexts
- [x] Remove `base` flag from Currency
- [x] Remove *.class parameters from Currency, CurrencyBundle and MoneyBundle (ref. #5001)
- [x] Write scenarios for switching currency
- [x] Handle changing currency
- [x] Handle unavailable previously selected currency
- [x] Handle unavailable default currency
- [x] Do not even use `ChannelContext` in **AdminBundle**
- [x] Use `ChannelContext` only in **ShopBundle**
- [x] Much more!
- [x] Wait for #5535 in order to remove 9672540 commit

### CODE REVIEW

File: features/currency/handling_disabled_currencies.feature [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        But currency "GBP" is disabled
+
+    Scenario: Not showing the disabled currency
+        When I browse that channel
+        Then it should not have ability to switch current currency to "GBP"
+
+    Scenario: Failing to browse channel with disabled default currency
+        Given currency "EUR" is disabled as well
+        When I try to browse that channel
+        Then I should not be able to do it
```
Comment (pjedrzejewski) [inline features/currency/handling_disabled_currencies.feature]: Then I should receive an error?

File: features/currency/handling_disabled_currencies.feature [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    Scenario: Browsing a channel with disabled default currency while using the other one
+        Given currency "EUR" is disabled as well
+        When I browse that channel while having "USD" as the current currency
+        Then it should have "USD" as the current currency
```
Comment (pjedrzejewski) [inline features/currency/handling_disabled_currencies.feature]: Then my currency should be "USD"

current currency sounds strange ;)

File: features/currency/handling_disabled_currencies.feature [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+    Background:
+        Given the store operates on a channel named "Web"
+        And that channel operates on "EUR", "USD" and "GBP" currencies
+        And it uses "EUR" currency by default
+        But currency "GBP" is disabled
+
+    Scenario: Not showing the disabled currency
+        When I browse that channel
+        Then it should not have ability to switch current currency to "GBP"
```
Comment (pjedrzejewski) [inline features/currency/handling_disabled_currencies.feature]: Then I should not be able to shop in "GBP"
Then I should not be able to pay in "GBP"
Then I should not be able to browse the store with "GBP" as currency

Something along these lines?

Comment (pamil) [inline features/currency/handling_disabled_currencies.feature]: `I should not be able to shop in "GBP"` sounds good

File: features/currency/handling_disabled_currencies.feature (hunk 2 above)
Comment (pamil) [inline features/currency/handling_disabled_currencies.feature]: According to: https://github.com/Sylius/Sylius/pull/5497/commits/b6dacb070297ccf70ec32fcd80bbfb78cb3d97c4#r70439049

`I should shop in "USD"`?

File: features/currency/handling_different_currencies_on_multiple_channels.feature [hunk 4]
```diff
@@ -0,0 +1,30 @@
+@currencies
+Feature: Handling different currencies on multiple channels
+    In order to see right prices
+    As a Customer
+    I want to browse channels only with valid currency
+
+    Background:
+        Given the store operates on a channel named "Web"
+        And that channel allows for "EUR", "USD" and "GBP" currencies
```
Comment (lchrusciel) [inline features/currency/handling_different_currencies_on_multiple_channels.feature]: `And that channel allows for "EUR", "USD" and "GBP" currencies` -> `And that channel allows for paying in euro, american dollars and british pounds currencies`?

File: src/Sylius/Bundle/CoreBundle/spec/Templating/Helper/PriceHelperSpec.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+use Sylius\Bundle\CoreBundle\Templating\Helper\PriceHelper;
+use Sylius\Bundle\MoneyBundle\Templating\Helper\PriceHelperInterface;
+use Sylius\Component\Currency\Context\CurrencyContextInterface;
+use Sylius\Component\Currency\Model\CurrencyInterface;
+use Symfony\Component\Templating\Helper\HelperInterface;
+
+/**
+ * @mixin PriceHelper
+ *
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/spec/Templating/Helper/PriceHelperSpec.php]: Good job, Łukasz Chruściel!

Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/spec/Templating/Helper/PriceHelperSpec.php]: XD

File: src/Sylius/Bundle/CoreBundle/spec/Templating/Helper/MoneyHelperSpec.php [hunk 6]
```diff
@@ -12,63 +12,57 @@
...
 use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Symfony\Component\Templating\Helper\HelperInterface;
 
 /**
  * @mixin MoneyHelper
  *
  * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
  */
-class MoneyHelperSpec extends ObjectBehavior
+final class MoneyHelperSpec extends ObjectBehavior
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/spec/Templating/Helper/MoneyHelperSpec.php]: I think it would be nice to introduce final to rest of spec classes.

File: features/currency/handling_different_currencies_on_multiple_channels.feature [hunk 7]
```diff
@@ -0,0 +1,30 @@
+@currencies
+Feature: Handling different currencies on multiple channels
+    In order to see right prices
+    As a Customer
+    I want to browse channels only with valid currency
```
Comment (TheMadeleine) [inline features/currency/handling_different_currencies_on_multiple_channels.feature]: ``I want to browse channels only with a valid currency``

File: features/currency/handling_different_currencies_on_multiple_channels.feature (hunk 4 above)
Comment (TheMadeleine) [inline features/currency/handling_different_currencies_on_multiple_channels.feature]: ``And that channel allows to pay with "EUR", "USD" and "GBP" currencies`` ?

File: features/currency/handling_different_currencies_on_multiple_channels.feature [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+    Background:
+        Given the store operates on a channel named "Web"
+        And that channel allows for "EUR", "USD" and "GBP" currencies
+        And it uses "EUR" currency by default
```
Comment (TheMadeleine) [inline features/currency/handling_different_currencies_on_multiple_channels.feature]: ``And it uses the "EUR" currency by default``

File: features/currency/handling_different_currencies_on_multiple_channels.feature [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        And the store operates on another channel named "Mobile"
+        And that channel allows for "USD" and "GBP" currencies
+        And it uses "USD" currency by default
+
+    Scenario: Showing currencies only from the current channel
+        When I browse "Mobile" channel
```
Comment (TheMadeleine) [inline features/currency/handling_different_currencies_on_multiple_channels.feature]: ``When I browse the "Mobile" channel``

File: src/Sylius/Bundle/CoreBundle/spec/Templating/Helper/MoneyHelperSpec.php (hunk 6 above)
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/spec/Templating/Helper/MoneyHelperSpec.php]: Will do!

File: features/currency/handling_different_currencies_on_multiple_channels.feature [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        Then I should shop using "USD" currency
+        And I should be able to shop using "GBP" currency
+        And I should not be able to shop using "EUR" currency
```
Comment (TheMadeleine) [inline features/currency/handling_different_currencies_on_multiple_channels.feature]: **the** "xxx" currency - in all steps in this file :)

File: features/currency/handling_different_currencies_on_multiple_channels.feature [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+    Scenario: Browsing channels using their default currencies
+        When I browse "Web" channel
+        And I start browsing "Mobile" channel
```
Comment (TheMadeleine) [inline features/currency/handling_different_currencies_on_multiple_channels.feature]: ``And I start browsing the "Mobile" channel``

File: features/currency/handling_different_currencies_on_multiple_channels.feature [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        Then I should shop using "USD" currency
+
+    Scenario: Switching currency applies only to the current channel
```
Comment (TheMadeleine) [inline features/currency/handling_different_currencies_on_multiple_channels.feature]: Switching **a** currency applies only to the current channel

File: features/currency/handling_different_currencies_on_multiple_channels.feature [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+        When I browse "Web" channel
+        And I switch the current currency to "GBP" currency
+        And I start browsing "Mobile" channel
+        Then I should still shop using "USD" currency
```
Comment (TheMadeleine) [inline features/currency/handling_different_currencies_on_multiple_channels.feature]: **the** "USD" currency

File: features/currency/handling_disabled_currencies.feature [hunk 14]
```diff
@@ -0,0 +1,26 @@
+@currencies
+Feature: Handling disabled currencies
+    In order to see right prices
+    As a Customer
+    I want to browse channels only with valid currency
```
Comment (TheMadeleine) [inline features/currency/handling_disabled_currencies.feature]: ``I want to browse channels with a valid currency only``

File: src/Sylius/Bundle/UiBundle/Resources/views/Form/theme.html.twig [hunk 15]
```diff
@@ -40,7 +40,7 @@
 
 {% block money_widget -%}
     <div class="ui labeled input">
-        <div class="ui label">{{ sylius_currency_symbol() }}</div>
+        <div class="ui label">{{ shopperContext.currencyCode|sylius_currency_symbol }}</div>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/UiBundle/Resources/views/Form/theme.html.twig]: shopper? chopper? harley? Why don't you use a `shop` prefix instead of `shopper`?

File: src/Sylius/Bundle/WebBundle/Menu/FrontendMenuBuilder.php [hunk 16]
```diff
@@ -58,9 +58,9 @@ class FrontendMenuBuilder extends MenuBuilder
     /**
      * Currency converter helper.
```
Comment (lchrusciel) [inline src/Sylius/Bundle/WebBundle/Menu/FrontendMenuBuilder.php]: Redundant

File: src/Sylius/Bundle/UiBundle/Resources/views/Form/theme.html.twig (hunk 15 above)
Comment (pamil) [inline src/Sylius/Bundle/UiBundle/Resources/views/Form/theme.html.twig]: https://github.com/Sylius/Sylius/blob/master/src/Sylius/Component/Core/Context/ShopperContext.php

File: src/Sylius/Component/Currency/Context/ImmutableCurrencyContext.php [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Currency\Context;
+
+use Sylius\Component\Currency\Model\CurrencyInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+
+final class ImmutableCurrencyContext implements CurrencyContextInterface
```
Comment (lchrusciel) [inline src/Sylius/Component/Currency/Context/ImmutableCurrencyContext.php]: Missing author

File: features/currency/switching_current_currency.feature [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+        And that channel allows for "EUR" and "USD" currencies
+        And it uses "EUR" currency by default
+
+    Scenario: Showing the current currency
+        When I browse that channel
+        Then I should shop using "EUR" currency
+
+    Scenario: Showing the available currencies
+        When I browse that channel
+        Then I should be able to shop using "USD" currency
```
Comment (TheMadeleine) [inline features/currency/switching_current_currency.feature]: Isn't that step too general?

File: features/currency/handling_disabled_currencies.feature [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+        Given the store operates on a channel named "Web"
+        And that channel allows for "EUR", "USD" and "GBP" currencies
+        And it uses "EUR" currency by default
+        But currency "GBP" is disabled
+
+    Scenario: Not showing the disabled currency
+        When I browse that channel
+        Then I should not be able to shop using "GBP" currency
+
+    Scenario: Failing to browse channel with disabled default currency
```
Comment (TheMadeleine) [inline features/currency/handling_disabled_currencies.feature]: ``Failing to browse channel with a disabled default currency``

File: features/currency/handling_disabled_currencies.feature [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+        Given currency "EUR" is disabled as well
```
Comment (TheMadeleine) [inline features/currency/handling_disabled_currencies.feature]: ``Given the currency "EUR" is disabled as well``

File: features/currency/handling_disabled_currencies.feature [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+        When I try to browse that channel
+        Then I should receive an error
```
Comment (TheMadeleine) [inline features/currency/handling_disabled_currencies.feature]: is it enough? Shouldn't we specify what kind of error is that?

File: features/currency/handling_disabled_currencies.feature [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+
+    Scenario: Browsing a channel with disabled default currency while using the other one
```
Comment (TheMadeleine) [inline features/currency/handling_disabled_currencies.feature]: Scenario: Browsing a channel with the default currency disabled while using another one

File: src/Sylius/Component/Core/Model/OrderItem.php [hunk 23]
```diff
@@ -79,6 +79,14 @@ public function getTaxTotal()
     }
 
     /**
+     * {@inheritdoc}
+     */
+    public function getRegularTotal()
+    {
+        return $this->total - $this->getAdjustmentsTotalRecursively(AdjustmentInterface::ORDER_UNIT_PROMOTION_ADJUSTMENT);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/OrderItem.php]: That is wrong, this method should be removed.

File: src/Sylius/Bundle/ShopBundle/Resources/views/Cart/Summary/_item.html.twig [hunk 24]
```diff
@@ -40,12 +40,11 @@
...
     <td class="right aligned">
         {# Todo: Fix this mess... #}
-        {% set regular_total = item|sylius_regular_price %}
         {% if 0 != item.getAdjustmentsTotalRecursively('order_unit_promotion') %}
-            <span class="sylius-total" style="text-decoration: line-through">{{ regular_total|sylius_price }}</span>
+            <span class="sylius-total" style="text-decoration: line-through">{{ item.regularTotal|sylius_price }}</span>
             <span class="sylius-discounted-total">{{ item.total|sylius_price }}</span>
         {% else %}
-            <span class="sylius-total">{{ regular_total|sylius_price }}</span>
+            <span class="sylius-total">{{ item.regularTotal|sylius_price }}</span>
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Cart/Summary/_item.html.twig]: As you can see above `{# Todo: Fix this mess... #}` that piece of code is wrong. We should display `item.unitPrice` and `item.discountedUnitPrice` if they are different (see order details in admin).

This change will require few adjustments in current promotions scenarios. Maybe we could extract it as a separate PR?

File: src/Sylius/Bundle/UiBundle/Resources/views/Form/theme.html.twig (hunk 15 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/UiBundle/Resources/views/Form/theme.html.twig]: Shouldn't we use default currency here? Isn't this widget used in admin where we don't have shopper context?

File: src/Sylius/Bundle/MoneyBundle/DependencyInjection/SyliusMoneyExtension.php [hunk 25]
```diff
@@ -32,11 +30,27 @@ public function load(array $config, ContainerBuilder $container)
         $config = $this->processConfiguration($this->getConfiguration($config, $container), $config);
         $loader = new XmlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
 
-        $container->setParameter('sylius.money.locale', $config['locale']);
-        $container->setParameter('sylius.money.currency', $config['currency']);
+        $container->setParameter('sylius_money.locale', $config['locale']);
+        $container->setParameter('sylius_money.currency', $config['currency']);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/MoneyBundle/DependencyInjection/SyliusMoneyExtension.php]: Do we need to add `sylius_money.currency` parameter? Are there any drawbacks of simpyl using `currency` parameter?

Comment (pamil) [inline src/Sylius/Bundle/MoneyBundle/DependencyInjection/SyliusMoneyExtension.php]: We can't be sure that there are `%currency%` or `%locale%` parameters in the application that uses MoneyBundle.

Comment (michalmarcinkowski) [inline src/Sylius/Bundle/MoneyBundle/DependencyInjection/SyliusMoneyExtension.php]: Ok, so we're using this to be able to take advantage of bundle configuration validation, sounds good to me then 👍  :)

File: src/Sylius/Bundle/ShopBundle/Resources/views/Currency/selector.html.twig [hunk 26]
```diff
@@ -0,0 +1,13 @@
+<div class="ui pointing dropdown link item" id="currency-selector">
```
Comment (Arminek) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Currency/selector.html.twig]: ``id="sylius-currency-selector"`` WDYT?

Comment (pamil) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Currency/selector.html.twig]: CSS namespaces? Sounds fun! 

comic-adult-swim-this-is-fine-3o6UBpHgaXFDNAuttm

File: src/Sylius/Behat/Context/Setup/CurrencyContext.php [hunk 27]
```diff
@@ -122,6 +117,48 @@ public function theStoreHasCurrencyWithExchangeRate($currencyCode, $exchangeRate
     }
 
     /**
+     * @Given /^(that channel) allows to shop using the "([^"]+)" currency$/
+     * @Given /^(that channel) allows to shop using "([^"]+)" and "([^"]+)" currencies$/
+     * @Given /^(that channel) allows to shop using "([^"]+)", "([^"]+)" and "([^"]+)" currencies$/
+     */
+    public function thatChannelAllowsToShopUsingAndCurrencies(
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/CurrencyContext.php]: Why don't you create three separated methods without optional parameter?

File: src/Sylius/Behat/Page/Shop/HomePageInterface.php [hunk 28]
```diff
@@ -27,4 +27,16 @@ public function getContents();
...
+     * @return string
+     */
+    public function getActiveCurrency();
+
+    /**
+     * @return array
+     */
+    public function getAvailableCurrencies();
+
+    public function switchCurrency($currencyCode);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/Shop/HomePageInterface.php]: Missing docblock

File: src/Sylius/Component/Core/Currency/CurrencyStorage.php [hunk 29]
```diff
@@ -0,0 +1,64 @@
...
+
+        return $currencyCode;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    private function provideKey(ChannelInterface $channel)
+    {
+        return sprintf('_currency_%s', $channel->getCode());
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Currency/CurrencyStorage.php]: I guess `_currency_ . $channel->getCode()` would be enough here.

File: src/Sylius/Bundle/MoneyBundle/composer.json [hunk 30]
```diff
@@ -28,6 +28,7 @@
         "webmozart/assert": "^1.0"
     },
     "require-dev": {
+        "sylius/currency-bundle": "^1.0",
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/MoneyBundle/composer.json]: Do we need it?

Comment (pamil) [inline src/Sylius/Bundle/MoneyBundle/composer.json]: In `require-dev` and `suggests` - yes :)

Comment (michalmarcinkowski) [inline src/Sylius/Bundle/MoneyBundle/composer.json]: `suggests` fine, but why in `require-dev`?

Comment (pamil) [inline src/Sylius/Bundle/MoneyBundle/composer.json]: We've got tests for MoneyBundle's integration with CurrencyBundle.

File: src/Sylius/Behat/Context/Setup/CurrencyContext.php (hunk 27 above)
Comment (pamil) [inline src/Sylius/Behat/Context/Setup/CurrencyContext.php]: We use the same approach for countries + it's less work :)

File: src/Sylius/Bundle/WebBundle/Menu/FrontendMenuBuilder.php (hunk 16 above)
Comment (pamil) [inline src/Sylius/Bundle/WebBundle/Menu/FrontendMenuBuilder.php]: Nothing to worry about, WebBundle is going to die soon.

File: src/Sylius/Bundle/CoreBundle/spec/Templating/Helper/MoneyHelperSpec.php (hunk 6 above)
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/spec/Templating/Helper/MoneyHelperSpec.php]: Done in #5541.

File: src/Sylius/Component/Core/Provider/ChannelAwareCurrencyProvider.php [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+use Sylius\Component\Channel\Context\ChannelNotFoundException;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Currency\Context\CurrencyNotFoundException;
+use Sylius\Component\Currency\Model\CurrencyInterface;
+use Sylius\Component\Currency\Provider\CurrencyProviderInterface;
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+final class ChannelAwareCurrencyProvider implements CurrencyProviderInterface
```
Comment (Arminek) [inline src/Sylius/Component/Core/Provider/ChannelAwareCurrencyProvider.php]: ``ChannelBasedCurrencyProvider`` WDYT?

Comment (pamil) [inline src/Sylius/Component/Core/Provider/ChannelAwareCurrencyProvider.php]: Yeah, but probably in another PR just after this one.

File: src/Sylius/Behat/Context/Setup/CurrencyContext.php (hunk 27 above)
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/CurrencyContext.php]: But it has this nasty foreach. But it just a small semantic change. I would prefer to have them separated, but this is irrelevant.

**Comment by pjedrzejewski** [comment]
Thanks Kamil, nice work! This PR is definitely too big and touches to many parts - which makes it hard to review. Please work with @TheMadeleine on documenting the changes **before** we start redoing locales. I have few things to add to this architecture.
